### ISSUE TITLE
CI: CircleCI seems to use branch config file on manual merge

### ISSUE THREAD

**Issue by tylerjereddy** [scipy/scipy#20051]
Studying the doc build failure in gh-20033 (https://app.circleci.com/pipelines/github/scipy/scipy/24759/workflows/231eede0-469d-4160-891a-05395f2be7e2/jobs/83510), it looks like we do correctly get a merge of the feature branch used for the PR with `main` (we do this "manually" as a step beforehand), but CircleCI seems to use the exact `.circleci/config.yml` on the submitted branch, since the old `python dev.py --no-build doc -j2` is still used there, triggering warnings about parallel build support with the newer sphinx theme/setup.

Intuitively, I can see how the chicken and egg type problem might happen depending on how they are setup, so I'm not sure this is worth much energy, maybe one documentation line for maintainers checking failures I suppose, or just refer back to this issue in the rare cases it matters. It could get messy if that file is getting changed a lot on `main` AND gets changed in a PR as well, but that is rare I think.

**Comment by lucascolley** [comment]
Hey @larsoner, any chance you could check if this is still relevant?

**Comment by larsoner** [comment]
Sounds like the `.circleci/config` has already been read by CircleCI and is being executed step-by-step when the merge step in the job happens. Doing that can't affect the set of steps, commands, etc. that would be run (this would require some sort of reloading mechanism or something)... *unless* you contain your steps in scripts. Having a tiny `.circleci/run_tests.sh` or similar as the command, CircleCI would have run that and it would have been changed by that point by the merge step, because the new code/command would have been merged into the script before CircleCI gets around to running it. This is one reason why (DRY being the other) in a lot of steps I've moved toward running little scripts in jobs rather than having bare Python or bash code.

Incidentally, the same sort of thing happens with GHA as well, so the same tradeoffs exist there, too.

FWIW, scripts are one way to deal with this problem, another is the "merge main into this branch" button, which also would have fixed it.

So my 2c -- if this happens often enough for the merge-main-into-branch becomes annoying, encapsulate commands in scripts.

**Comment by lucascolley** [comment]
> This is one reason why (DRY being the other) in a lot of steps I've moved toward running little scripts in jobs rather than having bare Python or bash code.

Yeah, I've been pushing for use of Pixi tasks in CI for the same reason. https://github.com/scipy/scipy/pull/24336 made an attempt at that (not quite eliminating all bash yet), but unfortunately it makes the flaky benchmarks (that sometimes crash at 50%) much more flaky (crashing every time that I tried). We should really figure that out.

**Comment by lucascolley** [comment]
Unfortunately, `asv` is just such a pain to debug 😅

**Comment by lucascolley** [comment]
Perhaps worth throwing fable at it

**Comment by larsoner** [comment]
If you point me to a good issue or MWE I can (have fable) look soon 😄 Would be more efficient than asking if to try to find the problem as well (though it probably would do that, too)

**Comment by lucascolley** [comment]
> If you point me to a good issue or MWE I can (have fable) look soon 😄 Would be more efficient than asking if to try to find the problem as well (though it probably would do that, too)

I'll try to merge main today to demonstrate the problem in CI. Part of the problem may be that it is reproducible on CircleCI but not easily reproducible locally.

**Comment by larsoner** [comment]
If push comes to shove "rerun wth SSH" on CircleCI might help with debugging... we'll see!

**Comment by lucascolley** [comment]
alright, thanks for agreeing to dig into this Eric! Hopefully (or hopefully not!), https://app.circleci.com/pipelines/github/scipy/scipy/42407/workflows/423c98c4-5f1d-49ea-b3d4-af86aada8a03 will show the crash I mean.

**Comment by larsoner** [comment]
In the meantime, I think we should probably close this since I think this is pretty standard CI behavior, even if it's a bit awkward to resolve sometimes. We can move steps into scripts as conflicts arise?
